library_generation/test/unit_tests.py

```diff
@@ -475,6 +511,71 @@ def test_monorepo_postprocessing_valid_repository_success(self):
             actual=f"{repository_path}/gapic-libraries-bom/pom.xml",
         )
 
+    @parameterized.expand(
+        [
+            (
+                "feat: a commit message\nPiperOrigin-RevId: 123456",
+                [
+                    "BEGIN_NESTED_COMMIT",
+                    "feat: [example_library] a commit message",
+                    "PiperOrigin-RevId: 123456",
+                    "END_NESTED_COMMIT",
+                ],
+                True,
+            ),
+            (
+                "feat: a commit message\nPiperOrigin-RevId: 123456",
+                [
+                    "BEGIN_NESTED_COMMIT",
+                    "feat: a commit message",
+                    "PiperOrigin-RevId: 123456",
+                    "END_NESTED_COMMIT",
+                ],
+                False,
+            ),
+            (
+                "feat: a commit message\nfix: an another commit message\nPiperOrigin-RevId: 123456",
+                [
+                    "BEGIN_NESTED_COMMIT",
+                    "feat: [example_library] a commit message",
+                    "fix: [example_library] an another commit message",
+                    "PiperOrigin-RevId: 123456",
+                    "END_NESTED_COMMIT",
+                ],
+                True,
+            ),
+            (
+                "feat: a commit message\nfix: an another commit message\nPiperOrigin-RevId: 123456",
+                [
+                    "BEGIN_NESTED_COMMIT",
+                    "feat: a commit message",
+                    "fix: an another commit message",
+                    "PiperOrigin-RevId: 123456",
+                    "END_NESTED_COMMIT",
+                ],
+                False,
+            ),
+        ]
+    )
+    def test_format_commit_message_success(self, message, expected, is_monorepo):
```

Comment:
This parameterized test can verify multi-line commit message can be formatted correctly.

Reply:
I'm not sure parameterized tests is the best choice here, because they are not intuit to read. Another developer has to count the position of the input to know which parameter is it. 
In addition, it groups all the test cases together, which make us harder to troubleshoot the unit tests if one particular test case fails, see [internal doc](https://engdoc.corp.google.com/eng/doc/devguide/testing/unit/best-practices.md?cl=head&polyglot=java) for a few best practices for unit tests. 

Ideally, we should have multiple test cases like
```
test_format_commit_message_should_add_library_name_if_regex_matches():

test_format_commit_message_should_not_add_library_name_if_regex_does_not_match():

test_format_commit_message_should_not_add_library_name_if_monorepo():

...
...
...


```

Reply:
I think a parameterized test is suitable here because
1. We need common setup for a mock commit object.
2. All different tests are duplicated in most parts.

I found python tips about [parameterized tests](go/python-tips/011#repetitive-tests) saying we should parameterize test in this case.